### PR TITLE
Tracker style API

### PR DESCRIPTION
**Pull request by KDr2**

### CODE REVIEW

**Comment by yebai** [comment]
Looks good to me. Can we come up with a simple example, where `Tracker` and `ThArrays` each differentiate part of a function? For example `f(x,y) = sin(x) * cos(y)`, where `sin(x)` is handled by `Tracker` and `cos(y)` is handled by `ThArrays`, or the other way.

**Comment by KDr2** [comment]
Hi, @yebai 

This PR now adds two new modules: `ThAD` and `TrackerAD`.

- `ThAD` provides interfaces like `gradient`, `forward`, etc., like other AD packages, and these functions use the PyTorch backend for all computation and AD.
- `TrackerAD` tries to mix PyTorch and Tracker.jl for AD computing, here is a simple example: https://github.com/TuringLang/ThArrays.jl/pull/14/commits/ae0ca6c395bd7b6336ede27e010aab3a2de1ee57#diff-bb371fe00e4cf75b62831b4f247b9b5c. This module is in a POC stage, and has much to do to make it complete.

**Comment by yebai** [comment]
thanks, @KDr2. I'll review this soon.

File: Project.toml [hunk 1]
```diff
@@ -7,7 +7,9 @@ version = "0.1.1-dev"
 
 [deps]
 Libdl = "8f399da3-3557-5675-b5ff-fb832c97cbdb"
+MacroTools = "1914dd2f-81c6-5fcd-8719-6d5c9610ff09"
 Pkg = "44cfe95a-1eb2-52ea-b672-e2afdf69b78f"
+Tracker = "9f7883ad-71c0-57eb-9f7f-b5c9e6d3789c"
```
Comment (yebai) [inline Project.toml:12]: Let's make `Tracker` optional dependencies using `Requires.jl`.

File: test/th_tr_mixed.jl [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+@testset "Torch and Tracker Mixed Gradient" begin
+
+    Random.seed!(0);
+
+    a = rand(3, 2)
+    b = rand(3, 2)
+
+    @testset "Simple Mixed AD" begin
+        # all with th(PyTorch Backend)
+        f1(x, y) = sum(sin.(_th(x)) + sin.(_th(y)))
```
Comment (yebai) [inline test/th_tr_mixed.jl:17]: Maybe improve the API by selecting AD backend based on the type of `x` and `y`, for example:

```julia
# sin.(y) with th, everything else with tr
f1(x::TrackedArray, y::Tensor) = sum(sin.(x) + sin.(y))
```
In general, functions are dispatched based on argument types (Tensor or TrackedArray). When mixed types are passed to a function, `Tensors` are automatically promoted to `TrackArray` by default (or the other way).

File: src/tracker.jl [hunk 3]
```diff
@@ -0,0 +1,203 @@
+module TrackerAD
```
Comment (yebai) [inline src/tracker.jl:1]: maybe move this file to `compat/tracker.jl` and load it optimally with `Requires.jl`.

File: src/common-methods.jl [hunk 4]
```diff
@@ -48,3 +47,30 @@ ThC.eye(::Type{T}, n::Int64; dev::Device=CPU()) where T =
     ThC.eye(n, eltype_id(T), convert(Int, dev))
 ThC.eye(::Type{Tensor{T}}, n::Int64; dev::Device=CPU()) where T =
     ThC.eye(T, n, dev=dev)
+
+Base.sum(t::Tensor{T}) where T = ThC.sum(t, eltype_id(T))
+Base.view(t::Tensor{T}, I...) where T = error("Not implement yet.")
+Base.transpose(t::Tensor{T, 2}) where T = ThC.t(t)
+Base.adjoint(t::Tensor) = error("Not implement yet.")
+# LinearAlgebra.det(t::Tensor) = error("Not implement yet.")
```
Comment (yebai) [inline src/common-methods.jl:55]: Nice to have a list of missing functions/operators from Base!

File: test/th_tr_mixed.jl (hunk 2 above)
Comment (KDr2) [reply test/th_tr_mixed.jl:17]: This way have two problems:
- We can't tell in which way the user would like to track the `+` operation, and we have more nested operators in the function body, we have no way to switch them forth and back.
- Generally, we just pass Arrays to `f1`, passing `TrackedArray` or `Tensor` may not be such convenient.

**Comment by KDr2** [comment]
Hi Hong, could you keep the commit history (don't squash them) when merging this PR? Thanks.

File: Project.toml (hunk 1 above)
Comment (KDr2) [reply Project.toml]: Done.

**Pull request merged**